Attempt to correctly handle indirect struct fields

dalehamel opened the pull request:
This is my WIP attempt to fix https://github.com/iovisor/bpftrace/issues/257, in the hopes of eventually fixing https://github.com/iovisor/bpftrace/issues/245. This is not ready to be merged, I'm putting this up mostly to share my work so far in the hopes that a review will shine some light on the proper solution to this problem

The root issue is the first one, which is that members of anonymous structs are not recognized. This is because of two problems:

1. These indirect fields are never recognized as fields of the struct in the first place
2. The offsets of these indirect fields would be incorrectly calculated if they were recognized, as they are taken relative to the field's direct parent (an anonymous struct or union) rather than to the struct whom they are indirect members of.

To fix problem one:

LLVM refers to members of an anonymous struct as "indirect fields". Indirect fields are not detected by the present implementation of struct parsing in bpftrace, as their parent will be an anonymous struct, so they will not be added as to the list of fields in the `structs_` hash. To fix this, I created `parse_anonymous_struct_name`, which will derive the name of the struct for on whom the indirect field exists. The existing solution is fragile, and I know already of some cases where it doesn't work. I believe however that this solution could be iterated upon.

To fix problem two:

I read through LLVM docs and source code, and found that the function `clang_Type_getOffsetOf` can be used to correctly determine the offsets of indirect fields, by manually specifying the struct type for whom they are indirect fields on (rather than their direct semantic parent). 

To do this, I walk the semantic ancestry of the cursor until I find the parent cursor, determine it's type, and take the offsets relative to that. In my test case, these offsets match what I would expect them to be when I compute them manually based on their sizes.

# Additional work

Before investing additional time in refining this solution, I'd like to see if this approach makes sense. If it does, then the major problems to overcome will be:

- `parse_anonymous_struct_name` will need to be more robust at converting to the name of the struct or union that actually contains the indirect fields.
- `indirect_field_offset` seems to produce the correct offsets, but when I attempt to `printf` the hex representation of the binary data stored at these offsets (by referencing the indirect field), the data that I get simply doesn't make sense, and doesn't match what similar scripts in iovisor/bcc produce. I suspect that even though these offsets are correct, there is a step somewhere that I've missed where retrieving the values of these indirect fields is incorrect. It might be that my approach is broken, and I need to find the address of the direct semantic parent for the indirect fields and compute offsets relative to that, but I've yet to confirm this.

@ajor sorry for yet another ping, but hopefully having some code will help to frame the issue described in https://github.com/iovisor/bpftrace/issues/257

ajor commented:
Thanks for looking into this! Your approach looks good to me, and seems to be working correctly when I try it on a simple example:
`
struct A
{
  int a;
  struct
  {
    int anon1;
    int anon2;
  };
  int b;
};
`
Do you have an example where it breaks?



Would it be possible to iterate up the cursor tree to find the first non-anon parent with:
`while(clang_Cursor_isAnonymous(parent))
{
  parent = clang_getCursorSemanticParent(parent);
 }`
rather than trying to extract the struct name from the anonymous type? I feel like that would be a bit nicer if it works.

dalehamel commented:
> Thanks for looking into this! 

No problem, thanks for creating an amazing tool @ajor 

> Your approach looks good to me, and seems to be working correctly when I try it on a simple example

Yeah for simple examples it certainly seems to work / match what I calculate in my head based on the sizes of the members, and it definitely seems to calculate the offset correctly, but the actual data that I try to read from these addresses doesn't seem cromulent (see below).

> Do you have an example where it breaks?

Yeah, so try to stay with me here (sorry it's a bit long, trying to give you lots of context).

I've been trying to re-implement BCC's [`tcpaccept`](https://github.com/iovisor/bcc/blob/master/tools/tcpaccept.py), here is my work-in-progress script, printing as hex since `inet_ntop` is not yet supported (if I can get this working and merged, that'll be my next project :) ):

```
#!/usr/bin/env bpftrace

/*
  traces tcp accepts
*/

#include <linux/tcp.h>
#include <net/sock.h>

BEGIN
{
  printf("Tracing tcp accepts. Hit Ctrl-C to end.\n");
  printf("%-8s %-16s %-8s %-6s %-6s %-6s\n", "TIME", "COMM", "SADDR", "SPORT", "DADDR", "DPORT")
}

kprobe:inet_csk_accept
{
  $sk = ((sock *) retval);
  time("%H:%M:%S ");
  printf("%-16s %04X:%02X %04X:%02X\n",
    comm,
    $sk->__sk_common.skc_rcv_saddr,
    $sk->__sk_common.skc_num,
    $sk->__sk_common.skc_daddr,
    $sk->__sk_common.skc_dport

  )
}
```

And here is the corresponding output:
```
TIME     COMM           SADDR:SPORT   DADDR:DPORT
23:18:23 nginx            89485533:89 B1443BE8:4156
23:18:23 nginx            89485533:89 B1443BE8:4156
23:18:23 ruby             89485533:89 B1443BE8:4156
23:18:23 ruby             89485533:89 B1443BE8:4156
23:18:23 ruby             89485533:89 B1443BE8:4156
23:18:23 ruby             89485533:89 B1443BE8:4156
23:18:23 ruby             89485533:89 B1443BE8:4156
```

This is the output of bcc's tcpaccept on the same machine around the same time:

```
PID    COMM         IP RADDR            LADDR            LPORT
2943646 nginx        4  10.228.225.102   10.228.34.120    8080
2943980 ruby         4  127.0.0.1        127.0.0.1        8000
817435 nginx        4  35.231.25.155    10.228.34.201    443
817436 nginx        4  35.229.81.75     10.228.34.201    443
2943646 nginx        4  10.229.18.113    10.228.34.120    8080
```

As you can see, the addresses from my version are pretty obviously incorrect - localhost (127.0.0.1 as 7f000001 / the correct endianness version of this) should be coming up and be pretty easy to spot in hexadecimal, yet I don't see it at all. In fact, all of the addresses are the same in this case! I tried another implementation, using the socket passed in as an argument to `inet_csk_accept`, and I do actually see different values, but they do not make sense as ip addresses as the CIDR values they correspond to are nonsense in our network topology.

To try and debug this further, I modified script, using arg0 (the socket being copied in `inet_csk_accept` instead:

```
#!/usr/bin/env bpftrace

/*
  traces tcp accepts
*/

#include <linux/tcp.h>
#include <net/sock.h>

BEGIN
{
  printf("Tracing tcp accepts. Hit Ctrl-C to end.\n");
  printf("%-8s %-16s %-8s %-6s %-6s %-6s\n", "TIME", "COMM", "SADDR", "SPORT", "DADDR", "DPORT")
}

kprobe:inet_csk_accept
{
  $sk = ((socket *) arg0)->sk;
  time("%H:%M:%S ");
  printf("%-16s %04X:%02X %04X:%02X\n",
    comm,
    $sk->__sk_common.skc_rcv_saddr,
    $sk->__sk_common.skc_num,
    $sk->__sk_common.skc_daddr,
    $sk->__sk_common.skc_dport

  )
}
```
And the corresponding output:

```
Tracing tcp accepts. Hit Ctrl-C to end.
TIME     COMM             SADDR    SPORT  DADDR  DPORT
23:26:57 nginx            FFFF8DA7:FFFF C7113B98:8DA8
23:26:57 nginx            FFFF8DA7:FFFF C7113B98:8DA8
23:26:57 nginx            FFFF8DAB:FFFF 350FCC98:8DAB
23:26:57 nginx            FFFF8DAB:FFFF 350FCC98:8DAB
23:26:57 ruby             FFFF8DB1:FFFF F12B8898:8DB8
23:26:57 ruby             FFFF8DB1:FFFF F12B8898:8DB8
23:26:57 ruby             FFFF8DB1:FFFF F12B8898:8DB8
23:26:57 ruby             FFFF8DB1:FFFF F12B8898:8DB8
```

While these addresses are at least different from one another, the `FFFF` in SADDR is clearly incorrect.

So, I suspect that while these offsets are calculated correctly from my debug prints (see https://elixir.bootlin.com/linux/v4.14/source/include/net/sock.h#L154 for reference for how this lines up with the source code):

```
Parsed: sock_common ident: skc_addrpair ptype: union sock_common::(anonymous at include/net/sock.h:155:2) offset: 0 size: 8
Parsed: sock_common ident: skc_daddr ptype: struct sock_common::(anonymous at include/net/sock.h:157:3) offset: 0 size: 4
Parsed: sock_common ident: skc_rcv_saddr ptype: struct sock_common::(anonymous at include/net/sock.h:157:3) offset: 4 size: 4
Parsed: sock_common ident: skc_hash ptype: union sock_common::(anonymous at include/net/sock.h:162:2) offset: 8 size: 4
Parsed: sock_common ident: skc_u16hashes ptype: union sock_common::(anonymous at include/net/sock.h:162:2) offset: 8 size: 4
Parsed: sock_common ident: skc_portpair ptype: union sock_common::(anonymous at include/net/sock.h:167:2) offset: 12 size: 4
Parsed: sock_common ident: skc_dport ptype: struct sock_common::(anonymous at include/net/sock.h:169:3) offset: 12 size: 2
Parsed: sock_common ident: skc_num ptype: struct sock_common::(anonymous at include/net/sock.h:169:3) offset: 14 size: 2
Parsed: sock_common ident: skc_family ptype: struct sock_common offset: 16 size: 2
```

that it is likely that the memory accessed from the pointer offset is incorrect in some way.

My exploration as I try to debug this further has lead me into the [llvm codegen](https://github.com/iovisor/bpftrace/blob/master/src/ast/codegen_llvm.cpp#L868-L874) for visiting FieldAccess objects. I have been trying to determine how `expr_` is set in this situation using `gdb`, but it's tricky to reason about this. `expr_` seems to be the base address of the struct for which these offsets are applied, which is why me investigation has taken me here.

As you wrote that code, I'm hopeful that you might have some insights as to why these values seem to be incorrect, or where I might look next.

It would be really helpful for me to be able to figure out how `expr_` is calculated for getting the base address of the struct here, as I would like to try setting that to the base-address of the anonymous struct that actually holds these members. I am curious to see if this + the offset relative to the direct parent comes up with the same address as my approach here. It could be that alignment, padding, or something I haven't considered makes my current approach just fundamentally invalid.

Or, perhaps there is something I'm missing about bcc's tcpaccept that is the "magic sauce" that makes it work here.

I'm really hoping to re-implement a number of the `tcp*` tools from bcc, and I think getting this functionality working is the critical missing piece right now.

> Would it be possible to iterate up the cursor tree to find the first non-anon parent ... rather than trying to extract the struct name from the anonymous type?

Yes something like this should work, the do-while loop you see there was me recycling some of my debugging code, I'll check for a simpler (and hopefully more robust) approach than the string parsing. I hadn't put more effort into this since the accessing the data at the fields for whom the offsets I've calculated didn't seem to make sense as above. 

If I can figure out how to get data that makes sense, I'll certainly figure out a better way to calculate the parent struct's name, but as it is now this functionality is probably misleading.

dalehamel commented:
> Good news for your first example - just need to change the kprobe to kretprobe and it all works! 

Oh my goodness is my face red - was so deep into it I missed the obvious. Thanks for pointing this out. I just need to type `skc_family` and I should be able to get this output to match pretty easily! 

> I don't think retval actually has anything useful in it on non-ret probes. If that's true then we should disallow it to avoid confusion like this in the future.

Yeah, incidentally would be nice to catch this and throw an error, I can see this confusion happening a lot through copy-pasta.

> I've not got your second one working but looking at the generated LLVM IR (using the -d flag), bpftrace seems to have the correct offsets for each of the fields it's looking for.

I wonder if if the state of the socket when I am trying to probe it is just different and isn't valuable to read? The original function I was trying to probe in https://github.com/iovisor/bpftrace/issues/245 was `tcp_v4_conn_request`, and specifically I was trying to re-write [this systemtap script](https://github.com/cloudflare/cloudflare-blog/blob/master/2018-01-syn-floods/acceptq.stp), which seems to actually pull the information I'm looking for (the source and destination address and port) from the socket buffer `skb` rather than from these struct fields.

I don't have systemtap set up, but it would be interesting to see if I get the same trying to use those fields instead of parsing the socket buffer. If so, it's likely that those values actually *need* to be read from the socket buffer. If that is the case it would be nice to be able to use `memcpy` to be able to pull fields out of a buffer (reading the network headers from there), as systemtap seems to do, perhaps there would be some use for a `read(buffer_ptr, num_bytes)` to bpftrace if this is the case?

For now, I'll try and clean up this PR to get it ready to merge - what should I do in terms of writing tests here?

dalehamel commented on src/clang_parser.cpp:
```diff
@@ -225,7 +251,7 @@ void ClangParser::parse(ast::Program *program, StructMap &structs)
         if (clang_getCursorKind(c) == CXCursor_FieldDecl)
         {
-          auto struct_name = get_clang_string(clang_getCursorSpelling(parent));
+          auto struct_name = get_parent_struct_name(c);
```
to minimize the diff, I decided to replace the call here rather than duplicate the logic below.

For non-anonymous structs it should give the same results it always has.

dalehamel commented on src/clang_parser.cpp:
```diff
@@ -234,6 +260,9 @@ void ClangParser::parse(ast::Program *program, StructMap &structs)
           auto ptypesize = clang_Type_getSizeOf(ptype);
 
+          if(clang_Cursor_isAnonymous(parent))
+            offset = get_indirect_field_offset(c, struct_name);
```
perhaps this could be simplified by having a single function to calculate offsets, as I've done above for struct names?

dalehamel commented:
@ajor I've cleaned up the code a lot. I found the file `test/clang_parser.cpp`, I think I understand the testing methodology. I'll add a regression test for anonymous struct handling / indirect field handling.

I think once I've added that and it is passing, this should be ready for a final review.

dalehamel commented:
Ok, this is ready for review.

I've:

* [ ] Added tests for the new functionality
* [ ] Verified that no existing tests are broken by this (it looks like the codegen tests are failing on master too, and there's a warning about this in the README)
* [ ] Verified that these tests fail on master, pass on my branch
* [ ] Refactored the code as much as I could think to

tyroguru commented:
@dalehamel - just to clarify here. With your handling of anon structs here are you expecting your tcp accepts script that you provided above to produce non-hex output? I was just giving it a try and for me it still produces hex values but I wasn't sure if that was what you expected.

dalehamel commented:
@tyroguru yes I am working on a separate patch to fix #30 to support inet_ntop so the addresses can be printed properly, the hex is “good enough” for now.

As a workaround, I tried modifying my script to do bit shifting in order to read the addresses; but couldn’t figure out how to use a hex mask properly (hex literals are undocumented if they work at all).

This PR just fixes fetching these actual struct members, it doesn’t fix how ip addresses are printed.

tyroguru commented:
> @tyroguru yes I am working on a separate patch to fix #30 to support inet_ntop so the addresses can be printed properly, the hex is “good enough” for now.
> 
> As a workaround, I tried modifying my script to do bit shifting in order to read the addresses; but couldn’t figure out how to use a hex mask properly (hex literals are undocumented if they work at all).
> 
> This PR just fixes fetching these actual struct members, it doesn’t fix how ip addresses are printed.

@dalehamel - cool. sounds great. That's what I expected but I just wanted to make sure the behaviour I'm see was what you expected me to see!

ajor commented:
Looks good!

dalehamel commented:
fyi @tyroguru I figured out how to hack around the lack of `inet_ntop` for now https://github.com/iovisor/bpftrace/issues/30#issuecomment-441156241, see https://github.com/iovisor/bpftrace/issues/245#issuecomment-441157715 for a complete example
